### ISSUE TITLE
[Gemma4/Blackhole] gemma-4-12B decode diverges from step 1 onward after warmup-captured trace at TP=8 (nkv_local=1 doubled sliding block)

### ISSUE THREAD

**Issue by arginugaTT** [tenstorrent/tt-metal#56577]
### Component / Area

models (gemma4) — vLLM hybrid-KV decode / warmup trace capture

### Observed

On Blackhole at **TP=8**, `gemma-4-12B-it` decode diverges from the **second decode
step onward** when the decode trace was captured during warmup. The first decode
step is correct. Output is uncorrelated with the reference (PCC ~0.06), not merely
degraded.

Black-box symptom: a vLLM server for `gemma-4-12B-it` on P150x8 emits degenerate
text (`"The capital of France is"` → `"11"`; `r1_gpqa_diamond` scores 0 with every
response empty). It reproduces on the **plain `tt_transformers` baseline entry**,
so it is not specific to speculative decoding. The same server stack serves
`gemma-4-31B-it` correctly, and `text_demo_v2.py` for 12B is fully coherent — so
this is specific to the vLLM warmup + hybrid-KV path.

**Two separable defects fail this test. This issue is (B).**

**(A) All-zero warmup page tables corrupt decode step 0.** Warmup captures with
`page_table=zeros`, binding every layer to block 0. Patching the test to warm up
with valid block ids (`torch.arange`) moves step 0 from PCC -0.068 to **0.9995**.
The plugin's real `TTModelRunner.warmup` also warms up with an all-zero page
table, so this is not test-only. Mentioned only because isolating (B) needs it
patched; closely related to #46432 ("warmup no longer covers per-layer tables …
can lead to bad outputs").

**(B) Steps 1+ diverge regardless — this issue.** With (A) patched out:

```
step 0: PCC 0.9995  PASS
step 1: PCC 0.0599  FAIL   u_top=27 v_top=28
step 2: PCC 0.0672  FAIL   u_top=27 v_top=29
step 3: PCC 0.0634  FAIL   u_top=27 v_top=29
```

`reset_batch=True` is passed only at step 0 — the one step that is correct.

**What distinguishes 12B from 31B (measured SDPA decode geometry at 1x8):**

| sliding layer, vLLM model | `nkv_local` | `eff_block` | k_cache padded |
|---|---|---|---|
| **12B (fails)** | **1** | **128** | `[16, 1, 128, 256]` |
| 31B (passes) | 2 | 64 | `[16, 2, 64, 256]` |

12B has 8 KV heads → `nkv_local = 8 // 8 = 1` at TP=8; 31B has 16 → 2. Because
12B's sliding layers then carry half the bytes of the full-attention spec, the HMA
per-block byte unifier **doubles the sliding block_size 64 → 128**. 31B's sliding
layers already match full's byte size (2×256 == 1×512), so nothing doubles. 12B at
TP=8 appears to be the only config here taking the doubled sliding-block path; 12B
also passes at 1x1, where `nkv_local` is 8. Global layers are `kv_replicated`
(`nkv_local=1`) in *both* models, so the split is in the **sharded sliding** layers.

**Ruled out by measurement:**

- Checkpoint/config loading — `text_demo_v2.py` for 12B is coherent.
- Weight-key prefix / tying — `model.language_model.*`, `tie_word_embeddings` with
  no `lm_head` tensor: same structure as the working 31B.
- Bounded sliding KV — fails with `GEMMA4_BOUNDED_SLIDING_KV_CACHE=0` too.
- On-device sampling — the test compares logits; both models init the same
  force-argmax path.
- `layer_types` — 12B 40 sliding + 8 full, 31B 50 + 10; same structure.
- Hybrid-KV prefill layout — `test_full_model_parity_uniform_vs_vllm` passes at
  PCC 1.0, and the test's prefill argmax assert does not fire.
- Trace capture at TP=8 in general — `test_full_model_parity_decode_trace` at 1x8
  **passes** (PCC 1.0). Capturing on real inputs is fine; only warmup capture fails.
- Per-layer page-table lifetime — persistent tensor ids stable across the
  warmup→inference boundary (no realloc), H2D copies land, skipped copies are
  legitimately identical content.
- Frozen SDPA geometry constants — geometry byte-identical at capture and at every
  replay step.
- On-device token/position feedback — non-PLI sets
  `_tt_vllm_always_refresh_decode_trace_inputs=False` by design (#51186) and relies
  on device feedback + position plus_one. But `GEMMA4_ALWAYS_REFRESH_DECODE=1`
  (forces host restage every step) **still** gives step 0 PASS / steps 1-3 FAIL. It
  changes behaviour (`v_top` unsticks from a constant 3 to 28/29) without fixing
  it, so the path is involved but is not the fault.

**Remaining shape:** the warmup-captured trace, replayed at steps 1+, is wrong even
with correct host-restaged inputs, while step 0 is correct. Unchecked suspects:
persistent staging buffers / `embed_idx` in the bounded sliding merge
(`models/demos/gemma4/tt/attention/decode.py`), and what `reset_batch=True` does
differently at step 0 (eager vs replay).

### Expected

Decode after a warmup-captured trace should match the uniform-layout reference at
every step, not just the first — i.e. `test_full_model_parity_warmup_then_inference`
passes at 1x8 for `gemma-4-12B-it` as it already does for `gemma-4-31B-it`, and a
vLLM server for 12B on P150x8 produces coherent text.

### 1. Steps (exact commands)

```bash
export TT_VISIBLE_DEVICES=0,1,4,5,8,9,12,13   # a valid (1,8) column
export MESH_DEVICE=P150x8
export HF_MODEL=google/gemma-4-12B-it
pytest models/demos/gemma4/tests/unit/test_vllm_parity.py \
  -k "warmup_then_inference and no-pli and 1x8" -q
```

```
FAILED test_full_model_parity_warmup_then_inference[blackhole-no-pli-small-steps4-1x8]
FAILED test_full_model_parity_warmup_then_inference[blackhole-pli-small-steps4-1x8]
2 failed, 2 skipped, 95 deselected in 46.46s
```

Control — the identical test with `HF_MODEL=google/gemma-4-31B-it` **passes**
(PCC ~0.999). The `-small` variant builds a synthetic config but derives attention
geometry from `HF_MODEL`, so the failure tracks the real 12B-vs-31B split.

To see (B) in isolation, warm up with non-zero page tables, i.e. replace in
`test_full_model_parity_warmup_then_inference`:

```python
warmup_page_table_per_layer = [torch.zeros(1, warmup_num_blocks, dtype=torch.int32) for _ in range(num_layers)]
# with
warmup_page_table_per_layer = [torch.arange(warmup_num_blocks, dtype=torch.int32).reshape(1, warmup_num_blocks) for _ in range(num_layers)]
```

### 2. Input data / link or description

Stock `google/gemma-4-12B-it` weights from the HF cache. No custom inputs; the test
builds its own synthetic config and prompts.

### 3. Frequency

Deterministic — 100% reproducible, identical PCC values run to run.

### 1. Software Versions

Branch `arg/gemma4_dflash_v2` @ `189299820a6`, forked from `main` @ `60053f1fb`.
The gemma4 changes on that branch are serving-side (vLLM generator) and are not in
this path — the failure reproduces through the unit test with no vLLM involved.

### 2. Hardware Details

Blackhole Galaxy (32× P150), `bh-glx-exp-b03u14`; 8 chips driven as an emulated
`P150x8` (1,8) column. `MESH_DEVICE=P150x8`, `FABRIC_1D`, Ubuntu 22.04.5.

### Is this a regression?

Unknown — not bisected.

### Logs & Diagnostics

Failing assertion values are inline above. Note for anyone reproducing on a Galaxy
box: only the **first** mesh open per process succeeds here; every later open in
the same pytest session dies with `Fabric Router Sync: Timeout ... Device 4`. Run
one parameterization per process (`-k "... and 1x8"`), or later tests fail
spuriously and read as real failures.

### Impact

Blocks serving `gemma-4-12B-it` through vLLM on P150x8 entirely — output is
degenerate on both the baseline and speculative entries. Note that throughput
benchmarks do **not** catch this: a full 35-point sweep returns `rc=0` with
"Acceptance: PASS" while emitting garbage, because a throughput benchmark counts
tokens without reading them.

**Comment by arginugaTT** [comment]
### Fix direction, derived from the working reference

The metal demo is a known-good reference for this model (12B `text_demo_v2.py` is
fully coherent at batch-1 and batch-8), and the parity suite is that comparison
codified. Holding model, mesh, layout and inputs constant and varying **only what
the decode trace was captured against**:

| Decode trace captured against | 12B @ 1x8 |
|---|---|
| **real** first-decode inputs (`test_full_model_parity_decode_trace`) | **PASS**, PCC 1.0 (all 4 steps) |
| **warmup zeros** (`test_full_model_parity_warmup_then_inference`) | FAIL, PCC ~0.06 |

So the corruption is not in tracing at TP=8, nor in the hybrid-KV layout — it is
specifically in capturing the decode trace against synthetic warmup inputs.

**Candidate fixes, in order of evidence:**

1. **Defer decode-trace capture to the first real decode step** (what the demo
   effectively does). `decode_trace` passing all 4 steps is direct evidence this
   works for 12B at 1x8. Cost: the ~seconds-long capture moves into the first
   request, which is what warmup exists to avoid — a latency trade that needs a
   decision, not an obvious win.
2. **Warm up with representative, non-degenerate inputs.** Necessary but
   **not sufficient on its own**: replacing the all-zero warmup page tables with
   valid block ids moves step 0 from PCC -0.068 to 0.9995, but steps 1+ still fail
   (0.060 / 0.067 / 0.063). Fixes (A), leaves (B).

**Diagnostic that does NOT work — don't repeat it.** Booting the baseline 12B
server with `--override-tt-config '{"trace_mode":"none"}'` to strip the warmup
capture end-to-end reaches READY and then EngineCore dies on the first request.
Cause is in the plugin, not the model: `async_decode.py:345-348` disables async
when `trace_mode == "none"`, while the baseline entry runs `async_scheduling:
true`. Testing that combination needs async disabled too (a `vllm_arg`, not
reachable from `--override-tt-config`).

For reference, `GEMMA4_ALWAYS_REFRESH_DECODE=1` is also not a fix — see the
ruled-out list in the issue body. Its own comment notes host restage under async
lag re-processes the previous token and the platform then disables async, so it
trades incoherency for a regression on async decode.

**Comment by arginugaTT** [comment]
### Correction: the black-box symptom in the issue body is wrong

**The server-level evidence quoted above is a probe artifact and should be
disregarded.** `gemma-4-12B-it` serves coherent text on this stack. Probed
correctly, the plain `tt_transformers` P150x8 entry returns:

```
"What is 2+2?"                 -> "2 + 2 is 4."                     finish=stop
"What is the capital of France?" -> "The capital of France is Paris." finish=stop
thinking mode, max_tokens=2048 -> "2+2 is 4."                       finish=stop
```

My original probes were wrong in two independent ways:

1. I used `/v1/completions`, which bypasses the chat template. An
   instruction-tuned model handed a bare sentence degenerates — that is where
   `"The capital of France is" -> "11"` and `'1.1111._...'` came from. It says
   nothing about the model.
2. My chat probes capped `max_tokens` at 23-150. Thinking mode is on by
   default for these entries, so a small cap truncates *inside* an unterminated
   thinking block and the reasoning parser emits neither `content` nor
   `reasoning_content`. I read that empty response as corruption.

The same two mistakes made `gemma-4-31B-it` look broken as well; it is not.

### What still stands

The **unit-test** failure is unaffected by any of this, because it never goes
through HTTP:

```
HF_MODEL=google/gemma-4-12B-it  warmup_then_inference @ 1x8 -> FAIL, PCC ~0.009
HF_MODEL=google/gemma-4-31B-it  warmup_then_inference @ 1x8 -> PASS, PCC ~0.999
```

Same mesh, same harness, same parameterization. Also unaffected:
`decode_trace` @ 1x8 passes (PCC 1.0), so the fault is specific to capturing the
decode trace against warmup zeros; and the measured `nkv_local=1` /
doubled-sliding-block geometry difference between the two models.

So the defect this issue describes is real and model-specific, but its
**severity is not established**. What I can support is a numerical divergence
from the uniform-layout reference after a warmup-captured trace. What I cannot
support — and withdraw — is the claim that it renders 12B serving unusable.

The `r1_gpqa_diamond` score of 0 with empty responses is a separate open
question I have not re-examined under a correct probe; it used
`max_gen_toks=32768`, so it was not truncated by a small cap, but "empty
content" there may be the same parser behaviour rather than token corruption.

Apologies for the noise in the original report.
